### ISSUE TITLE
Server-side Metadata for SEO

### ISSUE THREAD

**Issue by levarcooper** [aspnet/JavaScriptServices#554]
The Angular2 Demo Application does not update the title page. It displays "Home" for each view. I used the SetTitle property as explained in the Angular2 documentation but I believe the function references the document. Running it with an isBrowser conditional inside of ngOnInit works for aesthetics but search engines wouldn't detect the changes. What's the best way to update the metadata server-side?

**Comment by MarkPieszak** [comment]
Unfortunately, at the moment you'll need to just create your own interface to handle this like you can see below (https://github.com/angular/universal/issues/309#issuecomment-251718156)

Eventually I believe the regular `Title` service will work for both once universal is moved into ng Core itself.

```
import { Injectable, Inject, ElementRef, Renderer } from '@angular/core'
import { DOCUMENT } from '@angular/platform-browser'

@Injectable()
export class InterfaceService {    
    constructor (@Inject(DOCUMENT) private document) {}

    setTitle (title: string) {
        this.document.title = title
    }
}
```

Hope that helps!

**Comment by SteveSandersonMS** [comment]
Thanks @MarkPieszak!

**Comment by levarcooper** [comment]
Thanks for the reply, Mark. This seems to be similar to the Title service in angular 2 (client-side). Perhaps there's a problem with my implementation? Here's an example of what I'm doing app.component.ts after importing the service to app.module.ts and adding it to the providers array:
`export class AppComponent implements OnInit {
    meta;
    constructor(private _router: Router, private _activatedRoute: ActivatedRoute, private _metaService: MetaService) { }

    ngOnInit() {
        //this._metaService.setTitle('Title Test');  /* doesn't work */
        if (isBrowser) {
            this._metaService.setTitle('Title Test');  /* works */
        }
    }
}`
The code above gets the job done... sort of. It works, but non-javascript browsers and more importantly, search crawlers don't see the change to the title page.

**Comment by MarkPieszak** [comment]
Hmm this might be an issue with the .NET integration (since we're not getting the whole document, but rather just `<app></app>`, let me take a look tomorrow and see if I can get back to you with more info! @levarcooper

**Comment by levarcooper** [comment]
Thanks @MarkPieszak

**Comment by MarkPieszak** [comment]
@levarcooper So this is currently broken, and I'm hoping we can get in a Universal 2.4 version in sometime before the attempted merge into Angular Core. 
I'll keep you posted!

**Comment by MarkPieszak** [comment]
@SteveSandersonMS Should we keep this one open for now? 
(The way above won't work anymore)
It's not possible to be achieved at the moment without some workarounds most devs wouldn't be aware of.

**Comment by levarcooper** [comment]
@MarkPieszak, do you have more info about the workarounds you eluded to above. I was considering trying to pass javascript variables into the viewdata. Are there any known approaches that allow angular to communicate with dotnet at runtime?

**Comment by MarkPieszak** [comment]
Are you talking about passing it from .NET over to the front-end? @levarcooper 

https://github.com/aspnet/JavaScriptServices/tree/dev/src/Microsoft.AspNetCore.SpaServices#3-supplying-javascript-code-to-perform-prerendering

**Comment by ccaymaz** [comment]
I am having similar issue. I am using 'angular/platform-browser' to update metatags. It is working well when the page after loaded. But SPAService cannot complete request. It returns an error. 

--Basically I have and seo.service.ts to update tags. 
--home page (not included seo.service)
--about page(included seo.service)

When i hit home page first, my page loaded well. After that, I clicked about title has been changed. Looks like my service is  working on client side. 

BUT, If I hit about page first, page crashed.

> An unhandled exception occurred while processing the request.
> Exception: Call to Node module failed with error: Error: Uncaught (in promise): Error: DI Error
> Error: DI Error

In about page DI is not working somehow.
